File: src/org/gjt/jclasslib/browser/config/classpath/ClasspathDirectoryEntry.java
/*
    This library is free software; you can redistribute it and/or
    modify it under the terms of the GNU General Public
    License as published by the Free Software Foundation; either
    version 2 of the license, or (at your option) any later version.
*/

package org.gjt.jclasslib.browser.config.classpath;

import javax.swing.tree.DefaultTreeModel;
import java.io.File;

/**
    Classpath entry for a directory.

    @author <a href="mailto:jclasslib@ej-technologies.com">Ingo Kegel</a>
    @version $Revision: 1.1 $ $Date: 2003-08-18 08:10:15 $
*/
public class ClasspathDirectoryEntry extends ClasspathEntry {

    public FindResult findClass(String className) {

        File file = getFile();
        if (file == null) {
            return null;
        }
        File classFile = new File(file, className.replace('.', '/') + ".class");
        if (classFile.exists() && classFile.canRead()) {
            FindResult findResult = new FindResult(this, classFile.getPath());
            return findResult;
        }

        return null;
    }

    public void mergeClassesIntoTree(DefaultTreeModel model, boolean reset) {

        File directory = getFile();
        if (directory == null) {
            return;
        }

        ClassTreeNode rootNode = (ClassTreeNode)model.getRoot();
        mergeDirectory(directory, rootNode, model, reset);

    }

    private void mergeDirectory(File directory, ClassTreeNode parentNode, DefaultTreeModel model, boolean reset) {

        File[] files = directory.listFiles();
        if (files == null) {
            return;
        }
        for (int i = 0; i < files.length; i++) {
            File file = files[i];
            if (file.isDirectory()) {
                ClassTreeNode directoryNode = addOrFindNode(file.getName(), parentNode, true, model, reset);
                mergeDirectory(file, directoryNode, model, reset);
                if ((directoryNode.getChildCount() == 0)) {
                    int deletionIndex = parentNode.getIndex(directoryNode);
                    parentNode.remove(directoryNode);
                    if (!reset) {
                        model.nodesWereRemoved(parentNode, new int[] {deletionIndex}, new Object[] {directoryNode});
                    }
                }
            } else if (file.getName().toLowerCase().endsWith(CLASSFILE_SUFFIX)) {
                addOrFindNode(stripClassSuffix(file.getName()), parentNode, false, model, reset);
            }
        }

    }

}

File: modules/browser/src/main/java/org/gjt/jclasslib/browser/config/classpath/ClasspathDirectoryEntry.kt
/*
    This library is free software; you can redistribute it and/or
    modify it under the terms of the GNU General Public
    License as published by the Free Software Foundation; either
    version 2 of the license, or (at your option) any later version.
*/

package org.gjt.jclasslib.browser.config.classpath

import java.io.File
import javax.swing.tree.DefaultTreeModel

class ClasspathDirectoryEntry : ClasspathEntry() {

    override fun findClass(className: String): FindResult? {
        val file = file ?: return null
        val classFile = File(file, className.replace('.', '/') + ".class")
        if (classFile.exists() && classFile.canRead()) {
            return FindResult(this, classFile.path)
        }
        return null
    }

    override fun mergeClassesIntoTree(model: DefaultTreeModel, reset: Boolean) {
        val directory = file ?: return
        val rootNode = model.root as ClassTreeNode
        mergeDirectory(directory, rootNode, model, reset)
    }

    private fun mergeDirectory(directory: File, parentNode: ClassTreeNode, model: DefaultTreeModel, reset: Boolean) {
        val files = directory.listFiles() ?: return
        files.forEach { file ->
            if (file.isDirectory) {
                val directoryNode = addOrFindNode(file.name, parentNode, true, model, reset)
                mergeDirectory(file, directoryNode, model, reset)
                if (directoryNode.childCount == 0) {
                    val deletionIndex = parentNode.getIndex(directoryNode)
                    parentNode.remove(directoryNode)
                    if (!reset) {
                        model.nodesWereRemoved(parentNode, intArrayOf(deletionIndex), arrayOf<Any>(directoryNode))
                    }
                }
            } else if (file.name.toLowerCase().endsWith(ClasspathEntry.CLASSFILE_SUFFIX)) {
                addOrFindNode(stripClassSuffix(file.name), parentNode, false, model, reset)
            }
        }

    }

}
